api/leaderboard.go

```diff
@@ -10,842 +10,761 @@
+
+	if in.LeaderboardIds == "" {
+		app.AddError()
+		return nil, status.New(codes.InvalidArgument,
+			"Leaderboard IDs are required using the 'leaderboardIds' querystring parameter").Err()
+	}
 
-		status := 404
-		rank := 0
-		err := WithSegment("Model", c, func() error {
-			var err error
-			lg.Debug("Getting rank.")
-			rank, err = app.Leaderboards.GetRank(c.StdContext(), leaderboardID, memberPublicID, order)
+	leaderboardIDs := strings.Split(in.LeaderboardIds, ",")
+	serializedScores := make([]*api.GetRankMultiLeaderboardsResponse_Member, len(leaderboardIDs))
 
+	err := withSegment("Model", ctx, func() error {
+		for i, leaderboardID := range leaderboardIDs {
+			lg.Debug("Getting member rank on leaderboard.", zap.String("leaderboard", leaderboardID))
+			member, err := app.Leaderboards.GetMember(ctx, leaderboardID, in.MemberPublicId, order, in.ScoreTTL)
 			if err != nil && strings.HasPrefix(err.Error(), notFoundError) {
 				lg.Error("Member not found.", zap.Error(err))
 				app.AddError()
-				status = 404
-				return fmt.Errorf("Member not found.")
+				return status.New(codes.NotFound, "Leaderboard not found or member not found in leaderboard.").Err()
 			} else if err != nil {
-				lg.Error("Getting rank failed.", zap.Error(err))
+				lg.Error("Getting member rank on leaderboard failed.", zap.Error(err))
 				app.AddError()
-				status = 500
 				return err
 			}
-			lg.Debug("Getting rank succeeded.")
-			return nil
-		})
-		if err != nil {
-			return FailWith(status, err.Error(), c)
+			lg.Debug("Getting member rank on leaderboard succeeded.")
+			serializedScores[i] = &api.GetRankMultiLeaderboardsResponse_Member{
+				LeaderboardID: leaderboardID,
+				Rank:          int32(member.Rank),
+				Score:         float64(member.Score),
+				ExpireAt:      int32(member.ExpireAt),
+			}
 		}
-
-		return SucceedWith(map[string]interface{}{
-			"publicID": memberPublicID,
-			"rank":     rank,
-		}, c)
+		return nil
+	})
+	if err != nil {
+		return nil, err
 	}
+
+	return &api.GetRankMultiLeaderboardsResponse{
+		Success: true,
+		Scores:  serializedScores,
+	}, nil
 }
 
-//GetMemberRankInManyLeaderboardsHandler returns the member rank in several leaderboards at once
-func GetMemberRankInManyLeaderboardsHandler(app *App) func(c echo.Context) error {
-	return func(c echo.Context) error {
-		memberPublicID := c.Param("memberPublicID")
-		lg := app.Logger.With(
-			zap.String("handler", "GetMemberRankInManyLeaderboardsHandler"),
-			zap.String("memberPublicID", memberPublicID),
-		)
+// GetAroundMember retrieves a list of member score and rank centered in the given member
+func (app *App) GetAroundMember(ctx context.Context, in *api.GetAroundMemberRequest) (*api.GetAroundMemberResponse, error) {
+	lg := app.Logger.With(
+		zap.String("handler", "GetAroundMember"),
+		zap.String("leaderboard", in.LeaderboardId),
+		zap.String("memberPublicID", in.MemberPublicId),
+	)
+
+	order := in.Order
+	if order == "" || (order != "asc" && order != "desc") {
```

Comment:
This snippet appears in multiple places, consider making a function out of it.

Reply:
Done.